### ISSUE TITLE
`KrausChannel` doesn't work for non-invertible Kraus operators

### ISSUE THREAD

**Issue by DiegoGM91** [qiboteam/qibo#967]
The current implementation (`qibo v0.1.15` and `qibojit v0.0.10`) of the channel `gates.KrausChannel` breaks when the Kraus operators are singular (i.e. non-invertible). For example, for the amplitude damping channel. Here a snippet of code to reproduce this issue:

```python
from qibo.models import Circuit
from qibo import gates
import numpy as np

# Initialize Circuit
c = Circuit(1, density_matrix=True)
# Prepare initial state 1/sqrt(2) * (|0> + |1>)
c.add(gates.H(0))
# Define amplitude damping Kraus operators
p = 1e-1 # noise strength
E_1 = np.array([[0, np.sqrt(p)], [0, 0]])
E_2 = np.array([[1, 0], [0, np.sqrt(1-p)]])
# Apply Kraus operators
c.add(gates.KrausChannel(0, [E_1, E_2]))
# Check E_1 and E_2 are valid Kraus operators
print(np.dot(np.conjugate(np.transpose(E_1)), E_1) + np.dot(np.conjugate(np.transpose(E_2)), E_2))
# Execute the circuit
c()
```

The error can be traced back to a call of the function `matrix = np.linalg.inv(gate.asmatrix(self))`.

This seems to me to be a bug, and not only a `NotImplementedError` for singular Kraus operators, since it could happen that the Kraus operators are non-singular yet non-unitary.
This is so because what is needed for the application of Kraus operators is the dagger (conjugate transpose), which in general is not equal to the inverse. 

I think that implementing this would also replace computing the inverse of unitary Kraus operators by "brute force" by just taking the dagger, which should be faster if the Kraus operators are large.

**Comment by renatomello** [comment]
@DiegoGM91 fix implemented in https://github.com/qiboteam/qibojit/pull/128

**Comment by renatomello** [comment]
@DiegoGM91 However, I don't think it's as simple as just replacing `inv` for `dagger` for one reason: `qibojit` uses `inv` in order to avoid creating multiple copies of the state, so it reverses the implementation of one operator in order to apply the next one in the original state. If `dagger` doesn't equate `inv` and `inv` doesn't even exist, them it can't reverse back the state. I think the least disruptive solution is to use pseudo-inverse instead.

**Comment by DiegoGM91** [comment]
Thanks for the quick response @renatomello! 

However, I don't think the pseudo-inverse is an option here, as this will not implement the desired action for singular Kraus operators. (neither the inverse or pseudo-inverse will for non-unitary invertible Kraus operators)

If implementing the adjoint is not simple, then I guess the solution is checking for unitarity of the Kraus operators and throw a `NotImplementedError` if they are not unitary.

Althought it would be great and very useful to have this functionality in `qibo`!!

**Comment by renatomello** [comment]
@DiegoGM91 I think there is another way to solve this. I just realized that we had the same issue when implementing the `ReadoutErrorChannel`. The solution was to default to creating a copy of the original state. Something similar can be done here.

**Comment by DiegoGM91** [comment]
That'd be great. Thanks!

**Comment by renatomello** [comment]
@DiegoGM91 solution implemented in #968 and merged to the branch in #963. Will be in the `master` branch then #963  and https://github.com/qiboteam/qibojit/pull/127 are merged.
